### DISCUSSION TITLE
Tracers for CDT-Cloud

### DISCUSSION THREAD

**Discussion by MatthewKhouzam** [eclipse-cdt-cloud/cdt-cloud#6] [General]
In order to complete the developer loop, a tracer recommendation is needed to complete CDT.cloud's troubleshooting needs.

Here are some basic requirements to discuss.

1. The tracer should output in a format that can be understood by the trace viewer
2. The tracer should be simple enough for people to use in their hello world applications
3. The tracer needs to be license compatible with CDT.cloud
4. The tracer should trace something relevant to the use case of CDT.cloud (e.g. local spans)
5. The tracer should work in C/C++
6. The tracer should work in Linux, Mac OS and Windows
7. The tracer's impact on system performance should be low within reason
8. The tracer should be thread aware/safe on it's own. (part of being simple to use)

Here is a rundown of some tracers on the market
* LTTng UST (Linux only, high performance, not simple to use, very powerful/flexible) could be recommended
* Clang -ftime-trace (interesting, checks all the boxes, but it not configurable)
* Perf tracing http://blog.vmsplice.net/2017/07/tracing-userspace-static-probes-with.html (Linux only, advantages of being perf-like negated by trace viewer)
* SystemTap (configurable, Linux only, very flexible, not simple for hello worlds)
* Opentelemetry - Wrong scope, no C support as far as I can tell
* MicroProfile - Looks promissing, under "unliscence" so incompatible. https://github.com/jonasmr/microprofile
* New implementation - Requires more work than it may be worth.

Any thoughts on this?

**Comment by MatthewKhouzam** [comment]
Update: https://discourse.llvm.org/t/profiling-full-builds-with-ftime-trace-and-without-ninja/6050/3

**Reply by MatthewKhouzam** [reply]
Re: Microprofile, it looks like Roblox uses it. I asssume it's a safe option too.

Re: re-inventing the wheel... I am really against this. I consider it to be a sign of poor research.

**Comment by MatthewKhouzam** [comment]
I have a code example: 
[trace-array-example.zip](https://github.com/eclipse-cdt-cloud/cdt-cloud/files/9933115/trace-array-example.zip)

It iterates through a large array (Larger than the L2 cache on most CPUs) and shows the effects of memory thrashing.

**Comment by MatthewKhouzam** [comment]
Finally, I went ahead and coded an example application with a tracer built in... @sdirix could you tell me if the pi pico has built in cache? 
[xyyx.zip](https://github.com/eclipse-cdt-cloud/cdt-cloud/files/10416524/xyyx.zip)
This program has a simple tracer that I would like users to steal with pride. ;)

**Comment by sdirix** [comment]
According to the data sheet it has "External Quad-SPI Flash with eXecute In
Place (XIP) and 16kB on-chip cache". I'm not sure whether that matches the
assumptions of the posted code but I can certainly try. I'll check whether
I can manage to get this running on the Pico. Thanks!

On Sat, Jan 14, 2023 at 5:30 AM Matthew Khouzam ***@***.***>
wrote:

> Finally, I went ahead and coded an example application with a tracer built
> in... @sdirix <https://github.com/sdirix> could you tell me if the pi
> pico has built in cache?
> xyyx.zip
> <https://github.com/eclipse-cdt-cloud/cdt-cloud/files/10416524/xyyx.zip>
> This program has a simple tracer that I would like users to steal with
> pride. ;)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-cdt-cloud/cdt-cloud/discussions/6#discussioncomment-4683298>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEU3YFDWQRPO3FEEBXYSH3WSITVDANCNFSM5LBL2ZXA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Stefan Dirix

Senior Software Architect
EclipseSource

Email: ***@***.***
Web: https://eclipsesource.com

EclipseSource GmbH
Messerschmittstrasse 4
80992 Muenchen

General Managers: Dr. Jonas Helming, Dr. Maximilian Koegel
Registered Office: Messerschmittstrasse 4, 80992 Muenchen,
Commercial Register Muenchen, HRB 191789

**Reply by MatthewKhouzam** [reply]
In that case it should work! :) I'll go buy a pico soon.

**Reply by MatthewKhouzam** [reply]
I now have a pico... writing files is weird. It will take some work.
